File: spc_up/services/match/rules.py
"""Deterministic transaction matching rules (CPF/CNPJ extraction)."""

from __future__ import annotations

import re
import uuid
from typing import Literal

from sqlalchemy import select
from sqlalchemy.orm import Session

from spc_up.config import settings
from spc_up.models.entities import (
    MatchEvidencia,
    Movimentacao,
    MovimentacaoStatus,
    PessoaFisica,
    PessoaJuridica,
)
from spc_up.services.confidence import DEFAULT_WEIGHTS, evaluate_movimentacao
from spc_up.services.normalize import normalize_cnpj, normalize_cpf

_CPF_PATTERN = re.compile(
    r"\b(?:\d{3}\.?\d{3}\.?\d{3}-?\d{2}|\d{11})\b",
)
_CNPJ_PATTERN = re.compile(
    r"\b(?:\d{2}\.?\d{3}\.?\d{3}/?\d{4}-?\d{2}|"
    r"[A-Za-z0-9]{2}\.?[A-Za-z0-9]{3}\.?[A-Za-z0-9]{3}/?[A-Za-z0-9]{4}-?\d{2})\b",
)

STUB_PF_NOME = "DESCONHECIDO"
STUB_PJ_RAZAO = "DESCONHECIDA"


def _extract_document_candidates(descricao: str) -> list[tuple[Literal["CPF", "CNPJ"], str]]:
    seen: set[str] = set()
    candidates: list[tuple[Literal["CPF", "CNPJ"], str]] = []

    for match in _CNPJ_PATTERN.finditer(descricao):
        raw = match.group(0)
        try:
            normalized = normalize_cnpj(raw)
        except ValueError:
            continue
        if normalized not in seen:
            seen.add(normalized)
            candidates.append(("CNPJ", normalized))

    for match in _CPF_PATTERN.finditer(descricao):
        raw = match.group(0)
        try:
            normalized = normalize_cpf(raw)
        except ValueError:
            continue
        if normalized not in seen:
            seen.add(normalized)
            candidates.append(("CPF", normalized))

    return candidates


def _get_or_create_pessoa_fisica(session: Session, cpf: str) -> PessoaFisica:
    pessoa = session.scalar(select(PessoaFisica).where(PessoaFisica.cpf == cpf))
    if pessoa is not None:
        return pessoa

    pessoa = PessoaFisica(cpf=cpf, nome=STUB_PF_NOME)
    session.add(pessoa)
    session.flush()
    return pessoa


def _get_or_create_pessoa_juridica(session: Session, cnpj: str) -> PessoaJuridica:
    pessoa = session.scalar(select(PessoaJuridica).where(PessoaJuridica.cnpj == cnpj))
    if pessoa is not None:
        return pessoa

    pessoa = PessoaJuridica(cnpj=cnpj, razao_social=STUB_PJ_RAZAO)
    session.add(pessoa)
    session.flush()
    return pessoa


def _resolve_status(movimentacao: Movimentacao, score: float) -> MovimentacaoStatus:
    if score < settings.confianca_limiar_alta:
        return MovimentacaoStatus.PENDENTE_REVISAO

    pessoa_linked = movimentacao.pessoa_fisica_id is not None or movimentacao.pessoa_juridica_id is not None
    if pessoa_linked and not movimentacao.bloqueio_export:
        return MovimentacaoStatus.CONFIRMADO

    return MovimentacaoStatus.PENDENTE_REVISAO


def apply_deterministic_match(session: Session, movimentacao_id: uuid.UUID) -> Movimentacao:
    """Extract CPF/CNPJ from description, link pessoa, score and update status."""
    movimentacao = session.get(Movimentacao, movimentacao_id)
    if movimentacao is None:
        raise ValueError(f"Movimentacao {movimentacao_id} not found")

    movimentacao.evidencias.clear()

    cpfs: list[str] = []
    cnpjs: list[str] = []
    for doc_type, normalized in _extract_document_candidates(movimentacao.descricao_raw):
        if doc_type == "CPF":
            cpfs.append(normalized)
        else:
            cnpjs.append(normalized)

    if len(cpfs) > 1 or len(cnpjs) > 1 or (cpfs and cnpjs):
        movimentacao.evidencias.append(
            MatchEvidencia(
                tipo="CONFLITO_DOCUMENTO",
                peso=0.0,
                detalhe="Multiplos documentos encontrados na descricao",
            )
        )
        score = evaluate_movimentacao(movimentacao)
        movimentacao.status = _resolve_status(movimentacao, score)
        session.commit()
        session.refresh(movimentacao)
        return movimentacao

    if cpfs:
        cpf = cpfs[0]
        pessoa = _get_or_create_pessoa_fisica(session, cpf)
        movimentacao.pessoa_fisica_id = pessoa.id
        movimentacao.pessoa_juridica_id = None
        peso = DEFAULT_WEIGHTS["CPF_EXATO"]
        movimentacao.evidencias.append(
            MatchEvidencia(
                tipo="CPF_EXATO",
                peso=peso,
                detalhe=f"CPF {cpf} extraido da descricao",
            )
        )
    elif cnpjs:
        cnpj = cnpjs[0]
        pessoa = _get_or_create_pessoa_juridica(session, cnpj)
        movimentacao.pessoa_juridica_id = pessoa.id
        movimentacao.pessoa_fisica_id = None
        peso = DEFAULT_WEIGHTS["CPF_EXATO"]
        movimentacao.evidencias.append(
            MatchEvidencia(
                tipo="CNPJ_EXATO",
                peso=peso,
                detalhe=f"CNPJ {cnpj} extraido da descricao",
            )
        )

    score = evaluate_movimentacao(movimentacao)
    movimentacao.status = _resolve_status(movimentacao, score)
    session.commit()
    session.refresh(movimentacao)
    return movimentacao

File: packages/core/src/match/rules.ts
import type { Db, Movimentacao } from "@spc-up/db";
import { matchEvidencia, movimentacao } from "@spc-up/db";
import { eq } from "drizzle-orm";

import { evaluateMovimentacao } from "../confidence";
import { normalizeCnpj, normalizeCpf } from "../normalize";
import { MOVIMENTACAO_STATUS } from "../ingest/types";
import type { OrigemExtracaoV1 } from "../provenance/types";
import {
  type CadastroLinkTier,
  resolveCadastroLink,
} from "./cadastro-link";
import {
  CNPJ_PATTERN,
  CPF_PATTERN,
  findCnpjInDescricao,
  findCpfInDescricao,
  hasCpfInDescricao,
  stripDocumentsFromDescricao,
} from "./document-in-text";
import { structuredDocsFromOrigemExtracao } from "./structured-contraparte-docs";

export {
  findCnpjInDescricao,
  findCpfInDescricao,
  hasCpfInDescricao,
  stripDocumentsFromDescricao,
};

const DEFAULT_CONFIANCA_LIMITE_ALTA = 0.85;

type DocType = "CPF" | "CNPJ";

export function extractDocumentCandidates(
  descricao: string,
): Array<{ docType: DocType; normalized: string }> {
  const seen = new Set<string>();
  const candidates: Array<{ docType: DocType; normalized: string }> = [];

  for (const match of descricao.matchAll(CNPJ_PATTERN)) {
    const raw = match[0];
    try {
      const normalized = normalizeCnpj(raw);
      if (!seen.has(normalized)) {
        seen.add(normalized);
        candidates.push({ docType: "CNPJ", normalized });
      }
    } catch {
      continue;
    }
  }

  for (const match of descricao.matchAll(CPF_PATTERN)) {
    const raw = match[0];
    try {
      const normalized = normalizeCpf(raw);
      if (!seen.has(normalized)) {
        seen.add(normalized);
        candidates.push({ docType: "CPF", normalized });
      }
    } catch {
      continue;
    }
  }

  return candidates;
}

export function cleanNomeSugestao(desc: string, docString: string): string {
  let clean = desc.replace(docString, "").replace(/\b(?:CPF|CNPJ)\b/gi, "");
  clean = clean.replace(/\b(CRED PIX|DEB PIX|CRED TEV|DEB TEV|PIX TRANSF|PIX RECEBIDO|RECEBIDO EFETIVADO|PAGTO|PAGAMENTO|TED|DOC|DOC\/TED|DEPOSITO|SAQUE|TARIFA|SALDO|TRANSFERENCIA|TRANSF)\b/gi, "");
  clean = clean.replace(/[-|:,]/g, "").replace(/\s+/g, " ").trim();
  return clean.length >= 3 ? clean : "";
}

type MovimentacaoForEval = {
  confianca_global: number;
  bloqueio_export: boolean;
  pessoa_fisica_id?: string | null;
  pessoa_juridica_id?: string | null;
  spca?: null;
  evidencias: Array<{ tipo: string; peso: number | null }>;
};

function resolveStatusFromTier(
  tier: CadastroLinkTier,
  score: number,
  confiancaLimiteAlta: number,
  pessoaLinked: boolean,
  bloqueioExport: boolean,
): string {
  if (
    tier === "ALTA" &&
    score >= confiancaLimiteAlta &&
    pessoaLinked &&
    !bloqueioExport
  ) {
    return MOVIMENTACAO_STATUS.CONFIRMADO;
  }
  return MOVIMENTACAO_STATUS.PENDENTE_REVISAO;
}

export interface ApplyDeterministicMatchOptions {
  confiancaLimiteAlta?: number;
}

/** Extract CPF/CNPJ from description, link pessoa, score and update status. */
export async function applyDeterministicMatch(
  db: Db,
  movimentacaoId: string,
  options?: ApplyDeterministicMatchOptions,
): Promise<Movimentacao> {
  const confiancaLimiteAlta =
    options?.confiancaLimiteAlta ?? DEFAULT_CONFIANCA_LIMITE_ALTA;

  const rows = await db
    .select()
    .from(movimentacao)
    .where(eq(movimentacao.id, movimentacaoId))
    .limit(1);
  const current = rows[0];
  if (!current) {
    throw new Error(`Movimentacao ${movimentacaoId} not found`);
  }

  await db
    .delete(matchEvidencia)
    .where(eq(matchEvidencia.movimentacaoId, movimentacaoId));

  const origem = current.origemExtracao as OrigemExtracaoV1 | null | undefined;
  const structured = structuredDocsFromOrigemExtracao(origem);
  const link = await resolveCadastroLink(db, {
    cpf: structured.cpf,
    cnpj: structured.cnpj,
    remetenteDestinatario: current.remetenteDestinatario,
  });

  const pessoaFisicaId = link.pessoaFisicaId;
  const pessoaJuridicaId = link.pessoaJuridicaId;
  const evidencias = link.evidencias;

  if (evidencias.length > 0) {
    await db.insert(matchEvidencia).values(
      evidencias.map((ev) => ({
        movimentacaoId,
        tipo: ev.tipo,
        peso: ev.peso,
        detalhe: ev.detalhe,
      })),
    );
  }

  const movEval: MovimentacaoForEval = {
    confianca_global: current.confiancaGlobal,
    bloqueio_export: current.bloqueioExport,
    pessoa_fisica_id: pessoaFisicaId,
    pessoa_juridica_id: pessoaJuridicaId,
    spca: null,
    evidencias: evidencias.map((ev) => ({ tipo: ev.tipo, peso: ev.peso })),
  };

  const score = evaluateMovimentacao(movEval);
  const pessoaLinked =
    pessoaFisicaId != null || pessoaJuridicaId != null;
  const status = resolveStatusFromTier(
    link.tier,
    score,
    confiancaLimiteAlta,
    pessoaLinked,
    movEval.bloqueio_export,
  );

  const [updated] = await db
    .update(movimentacao)
    .set({
      pessoaFisicaId,
      pessoaJuridicaId,
      confiancaGlobal: movEval.confianca_global,
      bloqueioExport: movEval.bloqueio_export,
      status,
    })
    .where(eq(movimentacao.id, movimentacaoId))
    .returning();

  if (!updated) {
    throw new Error(`Failed to update movimentacao ${movimentacaoId}`);
  }

  return updated;
}
